logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FederationEventReceiver.kt

```diff
@@ -0,0 +1,230 @@
+        // remove pending and sent connections from federated users
+        connectionRepository.getConnections()
+            .map { it.firstOrNull() }
+            .onSuccess {
+                it?.forEach { conversationDetails ->
+                    if (conversationDetails is ConversationDetails.Connection
+                        && conversationDetails.otherUser?.id?.domain == event.domain
+                    ) {
+                        connectionRepository.deleteConnection(conversationDetails.conversationId)
+                    }
+                }
+            }
+
+        conversationRepository.getConversationsWithMembersWithBothDomains(event.domain, selfUserId.domain)
+            .onSuccess { conversationsWithMembers ->
+                // mark users as defederated to hold conversation history in oneOnOne conversations
+                conversationsWithMembers.oneOnOne.forEach { (conversationId, userIds) ->
+                    handleFederationDeleteEvent(conversationId, event.domain)
+                    userIds.filter { it.domain == event.domain }.forEach { userId ->
+                        userRepository.defederateUser(userId)
+                    }
+                }
+
+                // remove
+                conversationsWithMembers.group.forEach { (conversationId, userIds) ->
+                    handleFederationDeleteEvent(conversationId, event.domain)
+                    when (conversationId.domain) {
+                        // remove defederated users from self domain conversations
+                        selfUserId.domain ->
+                            removeMembersFromConversation(conversationId, userIds.filterNot { it.domain == selfUserId.domain })
+
+                        // remove self domain users from defederated domain conversations
+                        event.domain ->
+                            removeMembersFromConversation(conversationId, userIds.filter { it.domain == selfUserId.domain })
+
+                        // remove self domain and defederated domain users from not defederated and self domain conversations
+                        else -> removeMembersFromConversation(conversationId, userIds)
+                    }
+                }
+            }
+            .onSuccess {
+                kaliumLogger
+                    .logEventProcessing(
+                        EventLoggingStatus.SUCCESS,
+                        event
+                    )
+            }
+            .onFailure {
+                kaliumLogger
+                    .logEventProcessing(
+                        EventLoggingStatus.FAILURE,
+                        event,
+                        Pair("errorInfo", "$it")
+                    )
+            }
+    }
+
+    private suspend fun handleConnectionRemovedEvent(event: Event.Federation.ConnectionRemoved) =
+        withContext(dispatchers.io) {
+            val firstDomain = event.domains.first()
+            val secondDomain = event.domains.last()
```

Comment:
Are we sure here we are always going to receive 2 domains 😅 ? I don't think it is a problem, but maybe we can put here some logs, to monitor in case we receive an odd `> 2` case . What do you think ?

Reply:
Good point, I will log error and don't trigger any action when size will be different than 2